### ISSUE TITLE
retain messages are sent in parallel with CONNACK on re-connect

### ISSUE THREAD

**Issue by yentsun** [mcollina/aedes#80]
If a client re-connects to Aedes broker and has subscriptions with *retain* messages stored in session, it receives those messages in parallel with CONNACK. This probably violates MQTT 3.1.1 specs:
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.pdf (string 657)

> The first packet sent from the Server to the Client MUST be a CONNACK Packet 

While we don't have troubles with that with mqtt.js client:
```
1|⛭ watchm | 2017-01-22T07:11:53.450Z - debug: received message: <Buffer> over <topicName>
1|⛭ watchm | 2017-01-22T07:11:53.457Z - debug: connected: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0}
1|⛭ watchm | 2017-01-22T07:11:53.459Z - debug: received message: <Buffer> over <topicName>
...
proceeds normally
```

 Go (Paho) client responds with an error:
```
DEBUG 2017/01/19 18:36:11 client.go:178: [client]   Using MQTT 3.1.1 protocol
DEBUG 2017/01/19 18:36:11 client.go:359: [net]      connect started
ERROR 2017/01/19 18:36:11 client.go:375: [net]      received msg that was not CONNACK
ERROR 2017/01/19 18:36:11 client.go:376: PUBLISH: dup: false qos: 0 retain: true rLength: 170
topicName: <topicName> MessageID: 0
payload: <payload>
```

**Comment by mcollina** [comment]
wow, good catch! The problem arises because the clients sends a CONNECT and a SUBSCRIBE in parallel, and then the result is not deterministic, and it depends on your own hooks.

Would you mind in sending me a simple example using just mqtt-packet? Or maybe send a pull request to fix the issue?

Yes, MQTT.js is very lightweight in its checks on the spec from the server, there are plenty of implementations and not all are strict.

**Comment by yentsun** [comment]
> The problem arises because the clients sends a CONNECT and a SUBSCRIBE in parallel, and then the result is not deterministic, and it depends on your own hooks.

I guess not. Our js-client is essentially this:
```
function Agent (options) {

  if (!(this instanceof Agent)) {
    return new Agent(options)
  }

  const agent = mqtt.connect(url, {
    username: options.jwt,
    clean: false,
    clientId: options.deviceId
  });

  agent.on('error', function (error) {
    logger.error('connection error:', error.message);
  })
  agent.on('reconnect', function ()  {
    logger.debug('reconnecting...');
  });
  agent.on('close', function ()  {
    logger.debug('disconnected');
  });
  agent.on('connect', function (CONNACK)  {
    logger.debug('connected:', JSON.stringify(CONNACK));
  });
  agent.on('offline', function ()  {
    logger.debug('offline');
  });
  agent.on('message', function (topic, message, packet)  {
    logger.debug('received message:', message, 'over', topic);
  });

  this.agent = agent
  this.log = function (message) {
    agent.publish(someTopic, JSON.stringify({
        datetime: moment().format(),
        payload: {message: message}
      })
    )
  }

}

module.exports = Agent

```
It has no subscribes at all, but still receives subscriptions (along with CONNACK) from the previous set of experiments done with Paho client. So it seems like broker restores subscriptions, not the client, and this is the expected broker behavior, I guess.

Anyway we are thinking of a way to avoid retain topics permanent subscription so we wouldn't receive any messages on re-connect.

Eventually I could provide a pull request if I'm lucky to solve this.

**Comment by yentsun** [comment]
or maybe client initializes re-subscriptions in the background?

**Comment by mcollina** [comment]
> or maybe client initializes re-subscriptions in the background?

@yentsun yes! The problem is related to retain messages being set and offline subscriptions (clean=false).

See https://github.com/mcollina/aedes/blob/master/lib/handlers/connect.js#L109-L115 and https://github.com/mcollina/aedes/blob/master/lib/handlers/subscribe.js#L105-L119

**Comment by yentsun** [comment]
Ok, I see. True, the mqtt.js client does re-subscribe on re-connect: 
https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L192-L197

For now we solved this with  subscribe-receive-unsubscribe mechanic for 'retain' topics, so the client does not resubscribe/receive anything on reconnect, but I still will be trying to suggest a fix for Aedes. 

Thanks, Matteo!

**Comment by mcollina** [comment]
@yentsun the `'connect'` event is fired after CONNACK in MQTT.js
https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L757

The problem is that the server restores the subscriptions after connect if clean=false.
A fix would be highly welcomed.

**Comment by yentsun** [comment]
@mcollina Hi Matteo. Simply rearranging actions in `lib/handlers/connect.js`:

```udiff
 var connectActions = [
    authenticate,
    fetchSubs,
 -  restoreSubs,
    storeWill,
    doConnack,
 +  restoreSubs,
    emptyQueue
  ]
```
fixes the issue as we do CONNACK before restoring subscriptions (which if have retained messages in them violate the protocol) but breaks a couple of tests:
```
⨯ not clean and retain messages with QoS 1
  not ok 295 qos degraded to 0 for retained
    ---
      operator: equal
      expected: 0
      actual:   1
      at: DestroyableTransform.<anonymous> (/home/rollo/Projects/xshell/aedes/test/qos1.js:620:11)
    ...
  not ok 299 no error
    ---
      operator: fail
      at: Aedes.<anonymous> (/home/rollo/Projects/xshell/aedes/test/qos1.js:611:11)
    ...

```
which I can't easily fix. I actually don't understand why is QoS expected to degrade during the test.

Can you give me a hint here?

**Comment by mcollina** [comment]
@yentsun that test checks that retained messages are delivered _before_ offline messages. You should swap the check between the two `'data'`  events.

Would you mind to send a PR with a test for this case? We can work on it there together.

**Comment by yentsun** [comment]
@mcollina kk, I'll prepare the test case now

**Comment by yentsun** [comment]
@mcollina true - MQTT.js client seems to handle subscriptions *after* connack and I can't reproduce the behavior with it. Trying to add go-paho client to the test case...

**Comment by mcollina** [comment]
@yentsun you can use https://github.com/mqttjs/mqtt-connection, which is what aedes using internally for its tests: https://github.com/mcollina/aedes/blob/master/test/helper.js#L3-L8.

**Comment by yentsun** [comment]
@mcollina ok, I made a PR here https://github.com/mcollina/aedes/pull/94

**Comment by yentsun** [comment]
> @yentsun you can use https://github.com/mqttjs/mqtt-connection, which is what aedes using internally for its tests: https://github.com/mcollina/aedes/blob/master/test/helper.js#L3-L8.

@mcollina you think I'll be able to reproduce the behavior with mqtt-connection and not bother with Go client? will try...

**Comment by mcollina** [comment]
I think this might be fixed by #121. Feel free to reopen if it is still an issue.
